### PR TITLE
SAM tool important features

### LINKED ISSUE #646: Remaining SAM interface items
Task list for remaining items for SAM implementation in NimbusImage:

- [ ] **Enable "live preview".** It would be great to show a live preview of the mask that would be generated if the user clicks. It should just run SAM on the current point as the prompt. It doesn't need to process the mask at all—if it generates a complex mask, that is fine, the user will know that it may get simplified upon committing it to an annotation.
- [ ] **Progress bar for the encoder.** The encoder takes some time to compute, especially on older computers. It would be good to show some indicator on the screen or on the tool itself to show that it is doing the encoding. Perhaps also could be merged with this? https://github.com/Kitware/UPennContrast/issues/549
- [ ] **Interface for the tool.** I think there should be a toggle between "turbo mode" and "complex prompt mode". In turbo mode, you can shift click or shift drag, and it will automatically commit the annotation upon click/end of drag. That way, if someone just wants to quickly click on a bunch of objects, they can do so. In turbo mode, if you make an annotation, it gives you an "Oops, cancel!" link/button in the bottom left that allows you to undo the latest annotation quickly (lasts for 1 second). In complex prompt mode, it can allow for more complex prompts with multiple foreground and background prompt points/rectangles. In that mode, the existing buttons should be sufficient.
- [ ] **Simplify the polygon.** Give an optional "smoothing parameter" (default 1.5) to send to the algorithm to smooth out the resulting polygon so that it doesn't have quite so many points.

### PR DESCRIPTION
**Pull request by bruyeret**
Add SAM annotation tool features requested in #646 

Note: you need to rebuild the itk pipeline for SAM because of a small change (`npm run emscripten-build`)

Close #646

### CODE REVIEW

**Review by arjunrajlab — APPROVED** [review]
Tested it out, and it works great!

File: src/components/AnnotationViewer.vue [hunk 1]
```diff
@@ -272,29 +298,69 @@ export default class AnnotationViewer extends Vue {
...
+      this.annotationLayer.removeAnnotation(this.samLivePreviewAnnotation);
+      this.samLivePreviewAnnotation = null;
+    }
+
+    // Create the new annotation
+    const vertices = this.samLivePreviewOutput;
+    if (!vertices) {
+      return;
+    }
+    const style = {
```
Comment (luciemac) [inline src/components/AnnotationViewer.vue:329]: Don't you have the possibility to customise your annotation style?

File: src/pipelines/computePipeline.ts [hunk 2]
```diff
@@ -81,15 +80,46 @@ export class ComputeNode<
...
+  /**
+   * Create and return a callback function that will trigger recomputation of this node
+   * @param parentNode The parent node that will trigger a recomputation
+   * @returns The callback that needs to be called when the parentNode's output changes
+   */
+  private createComputeCallback(parentNode: ComputeNode<any, any>) {
+    let previousParentOutput = parentNode.output;
+    const computeCallback = () => {
+      const currentParentOutput = parentNode.output;
+      // Don't recompute when the parent output changes from NoOutput to NoOutput
```
Comment (luciemac) [inline src/pipelines/computePipeline.ts:102]: if it changes from NoOutput to NoOutput, then, it did not change :)

File: src/components/AnnotationViewer.vue (hunk 1 above)
Comment (bruyeret) [reply src/components/AnnotationViewer.vue:329]: Not yet but it could be possible in the future :D

File: src/pipelines/computePipeline.ts (hunk 2 above)
Comment (bruyeret) [reply src/pipelines/computePipeline.ts:102]: Good point!

File: src/pipelines/samPipeline.ts [hunk 3]
```diff
@@ -341,127 +356,255 @@ async function runItkPipeline({ masks }: IDecoderOutput): Promise<IItkOutput> {
...
+  const { encoderNodes, decoderNodes, previewNodes } = nodes.allNodes;
+  const computingMessageMap: [ComputeNode<any, any>, string][] = [
+    [encoderNodes.sessionNode, "Creating encoder"],
+    [encoderNodes.inferenceNode, "Encoding"],
+    [decoderNodes.sessionNode, "Creating decoder"],
+    [previewNodes.sessionNode, "Creating decoder preview"],
+  ];
+  const recomputeLoadingMessage = () => {
+    const messages: string[] = [];
+    for (const [node, message] of computingMessageMap) {
```
Comment (luciemac) [inline src/pipelines/samPipeline.ts:575]: Can't you use a `filter` here ?

**Comment by luciemac** [comment]
Then, LGTM

File: src/components/AnnotationViewer.vue (hunk 1 above)
Comment (luciemac) [reply src/components/AnnotationViewer.vue:329]: Don't know if it is useful for now, but maybe you can generalize this style, to make it easily customizable, what do you think?

**Pull request merged**